File: commands/blackjack.js
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const { colors } = require('../config.js');

class BlackjackGame {
    constructor() {
        this.deck = this.createDeck();
        this.playerHand = [];
        this.dealerHand = [];
    }

    createDeck() {
        const suits = ['♠', '♥', '♦', '♣'];
        const values = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
        const deck = [];
        
        for (const suit of suits) {
            for (const value of values) {
                deck.push({ suit, value });
            }
        }
        
        return this.shuffle(deck);
    }

    shuffle(deck) {
        for (let i = deck.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [deck[i], deck[j]] = [deck[j], deck[i]];
        }
        return deck;
    }

    drawCard() {
        return this.deck.pop();
    }

    calculateHand(hand) {
        let sum = 0;
        let aces = 0;

        for (const card of hand) {
            if (card.value === 'A') {
                aces += 1;
                sum += 11;
            } else if (['J', 'Q', 'K'].includes(card.value)) {
                sum += 10;
            } else {
                sum += parseInt(card.value);
            }
        }

        while (sum > 21 && aces > 0) {
            sum -= 10;
            aces -= 1;
        }

        return sum;
    }

    formatHand(hand, hidden = false) {
        if (hidden) {
            return `${hand[0].value}${hand[0].suit} | ?`;
        }
        return hand.map(card => `${card.value}${card.suit}`).join(' | ');
    }
}

module.exports = {
    data: new SlashCommandBuilder()
        .setName('blackjack')
        .setDescription('Jouer au Blackjack'),

    async execute(interaction) {
        const game = new BlackjackGame();
        
        // Distribution initiale
        game.playerHand.push(game.drawCard(), game.drawCard());
        game.dealerHand.push(game.drawCard(), game.drawCard());

        const createGameEmbed = (hideDealer = true, gameOver = false) => {
            const embed = new EmbedBuilder()
                .setColor(colors.primary)
                .setTitle('🎲 Blackjack')
                .addFields(
                    { name: '🎰 Croupier', value: game.formatHand(game.dealerHand, hideDealer) },
                    { name: '👤 Vous', value: game.formatHand(game.playerHand) }
                );

            if (gameOver) {
                const playerScore = game.calculateHand(game.playerHand);
                const dealerScore = game.calculateHand(game.dealerHand);
                
                let result;
                if (playerScore > 21) {
                    result = 'Vous avez dépassé 21! Perdu!';
                } else if (dealerScore > 21) {
                    result = 'Le croupier a dépassé 21! Gagné!';
                } else if (playerScore > dealerScore) {
                    result = 'Vous avez gagné!';
                } else if (playerScore < dealerScore) {
                    result = 'Le croupier gagne!';
                } else {
                    result = 'Égalité!';
                }
                
                embed.addFields({ name: 'Résultat', value: result });
            }

            return embed;
        };

        const response = await interaction.reply({
            embeds: [createGameEmbed()],
            components: [
                {
                    type: 1,
                    components: [
                        {
                            type: 2,
                            custom_id: 'hit',
                            label: 'Tirer',
                            style: 1,
                        },
                        {
                            type: 2,
                            custom_id: 'stand',
                            label: 'Rester',
                            style: 1,
                        }
                    ]
                }
            ],
            fetchReply: true
        });

        const collector = response.createMessageComponentCollector({
            time: 60000
        });

        collector.on('collect', async i => {
            if (i.user.id !== interaction.user.id) {
                return i.reply({ content: 'Ce n\'est pas votre partie!', ephemeral: true });
            }

            if (i.customId === 'hit') {
                game.playerHand.push(game.drawCard());
                const playerScore = game.calculateHand(game.playerHand);

                if (playerScore > 21) {
                    collector.stop();
                    await i.update({
                        embeds: [createGameEmbed(false, true)],
                        components: []
                    });
                } else {
                    await i.update({
                        embeds: [createGameEmbed()]
                    });
                }
            } else if (i.customId === 'stand') {
                let dealerScore = game.calculateHand(game.dealerHand);
                while (dealerScore < 17) {
                    game.dealerHand.push(game.drawCard());
                    dealerScore = game.calculateHand(game.dealerHand);
                }

                collector.stop();
                await i.update({
                    embeds: [createGameEmbed(false, true)],
                    components: []
                });
            }
        });

        collector.on('end', () => {
            if (!interaction.replied) {
                interaction.editReply({
                    components: []
                });
            }
        });
    },
};

File: cogs/blackjack.py
import discord
from discord import app_commands
from discord.ext import commands
import random
from typing import List, Dict

class Card:
    def __init__(self, suit: str, value: str):
        self.suit = suit
        self.value = value

    def __str__(self):
        return f"{self.value}{self.suit}"

class BlackjackGame:
    def __init__(self):
        self.deck = self.create_deck()
        self.player_hand: List[Card] = []
        self.dealer_hand: List[Card] = []

    def create_deck(self) -> List[Card]:
        suits = ['♠', '♥', '♦', '♣']
        values = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K']
        deck = [Card(suit, value) for suit in suits for value in values]
        random.shuffle(deck)
        return deck

    def draw_card(self) -> Card:
        return self.deck.pop()

    def calculate_hand(self, hand: List[Card]) -> int:
        total = 0
        aces = 0

        for card in hand:
            if card.value == 'A':
                aces += 1
                total += 11
            elif card.value in ['J', 'Q', 'K']:
                total += 10
            else:
                total += int(card.value)

        while total > 21 and aces > 0:
            total -= 10
            aces -= 1

        return total

    def format_hand(self, hand: List[Card], hide_dealer: bool = False) -> str:
        if hide_dealer:
            return f"{hand[0]} | ?"
        return " | ".join(str(card) for card in hand)

class BlackjackCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.games: Dict[int, BlackjackGame] = {}

    @app_commands.command(name="blackjack", description="Jouer au Blackjack")
    async def blackjack(self, interaction: discord.Interaction):
        if interaction.user.id in self.games:
            await interaction.response.send_message("Vous avez déjà une partie en cours!", ephemeral=True)
            return

        game = BlackjackGame()
        self.games[interaction.user.id] = game

        # Distribution initiale
        game.player_hand.extend([game.draw_card(), game.draw_card()])
        game.dealer_hand.extend([game.draw_card(), game.draw_card()])

        embed = discord.Embed(title="🎲 Blackjack", color=discord.Color.blue())
        embed.add_field(name="🎰 Croupier", value=game.format_hand(game.dealer_hand, True), inline=False)
        embed.add_field(name="👤 Vous", value=game.format_hand(game.player_hand), inline=False)

        # Création des boutons
        hit_button = discord.ui.Button(style=discord.ButtonStyle.primary, label="Tirer", custom_id="hit")
        stand_button = discord.ui.Button(style=discord.ButtonStyle.primary, label="Rester", custom_id="stand")
        
        view = discord.ui.View()
        view.add_item(hit_button)
        view.add_item(stand_button)

        async def hit_callback(interaction: discord.Interaction):
            game = self.games.get(interaction.user.id)
            if not game:
                return

            game.player_hand.append(game.draw_card())
            player_score = game.calculate_hand(game.player_hand)

            embed = discord.Embed(title="🎲 Blackjack", color=discord.Color.blue())
            embed.add_field(name="🎰 Croupier", value=game.format_hand(game.dealer_hand, True), inline=False)
            embed.add_field(name="👤 Vous", value=game.format_hand(game.player_hand), inline=False)

            if player_score > 21:
                embed.add_field(name="Résultat", value="Vous avez dépassé 21! Perdu!", inline=False)
                await interaction.response.edit_message(embed=embed, view=None)
                del self.games[interaction.user.id]
            else:
                await interaction.response.edit_message(embed=embed)

        async def stand_callback(interaction: discord.Interaction):
            game = self.games.get(interaction.user.id)
            if not game:
                return

            dealer_score = game.calculate_hand(game.dealer_hand)
            while dealer_score < 17:
                game.dealer_hand.append(game.draw_card())
                dealer_score = game.calculate_hand(game.dealer_hand)

            player_score = game.calculate_hand(game.player_hand)

            embed = discord.Embed(title="🎲 Blackjack", color=discord.Color.blue())
            embed.add_field(name="🎰 Croupier", value=game.format_hand(game.dealer_hand, False), inline=False)
            embed.add_field(name="👤 Vous", value=game.format_hand(game.player_hand), inline=False)

            if dealer_score > 21:
                result = "Le croupier a dépassé 21! Gagné!"
            elif player_score > dealer_score:
                result = "Vous avez gagné!"
            elif player_score < dealer_score:
                result = "Le croupier gagne!"
            else:
                result = "Égalité!"

            embed.add_field(name="Résultat", value=result, inline=False)
            await interaction.response.edit_message(embed=embed, view=None)
            del self.games[interaction.user.id]

        hit_button.callback = hit_callback
        stand_button.callback = stand_callback

        await interaction.response.send_message(embed=embed, view=view)

async def setup(bot):
    await bot.add_cog(BlackjackCog(bot))
